### ISSUE TITLE
[10.0.0] The ownCloud desktop client is temporarily removing most of my files locally when server gets into maintenance mode

### ISSUE THREAD

**Issue by fjf2002** [owncloud/client#5738]
### Expected behaviour
When switching the server into maintenance mode, the ownCloud desktop client should report it is not connected to the server and synchronization should stop/fail.

### Actual behaviour
When switching the server into maintenance mode, the ownCloud desktop client still reports it is connected to the server and deletes most of my local files.

I am using "local" external storage and I am synchronizing about 1,3G, if that does matter.

### Steps to reproduce
1. In the ownCloud desktop client, select "force Sync now"(?) ("Synchronisation jetzt erzwingen"). The desktop client will re-scan synchronized folders locally, which will take a while.

2. During the previous step, put the ownCloud server into maintenance mode, i. e. edit /var/www/owncloud/config/config.php and set maintenance to true.

3. The ownCloud desktop client will delete most local files.

### Server configuration
Operating system:
Ubuntu 16.10

Web server: 
Apache/2.4.18 (Ubuntu)

Database:
psql (PostgreSQL) 9.5.6

PHP version:
PHP 7.0.15-0ubuntu0.16.10.4

ownCloud version:
ownCloud	10.0.0 (stable)

Storage backend (external storage):
"Local" external storage

### Client configuration
Client version:
2.3.1 (build 6824)

Operating system:
Windows10

OS language:
german

Installation path of client:
C:\Program Files (x86)\ownCloud

**Comment by hodyroff** [comment]
Is this the same if you use the occ command? Editing config.php seems somewhat risky. https://doc.owncloud.org/server/10.0/admin_manual/configuration_server/occ_command.html
Are you doing something with the local storage at that time = disabling it in some way?

**Comment by ckamm** [comment]
My understanding of what should be happening is that the server's replies should become 503 once maintenance mode is activated. The client should then ignore files and folders with these replies. This needs a reproduction attempt.

**Comment by fjf2002** [comment]
@hodyroff: I just tested:
* It is the same when i use the occ command.
* Files that are really located in /var/www/owncloud/data/myuser/files donot seem to be affected. Only my  "local" external storage.
* Very strangely the owncloud desktop client eventually complains about not being able to synchronize one certain file that was removed - on both server and client - days ago(!) - with the server still being in maintenance mode(!)

**Comment by ckamm** [comment]
I can reproduce: I've set up a "local" external storage. When I switched the server to maintenance mode, the client deleted the local copy of the external storage folder. Investigating.

**Comment by ckamm** [comment]
@PVince81 @guruz It seems to be that maintenance mode does not set the new dav endpoint to 503!

```
curl 'localhost/remote.php/webdav' -X PROPFIND -u ... -v
-> 503
curl 'localhost/remote.php/dav/files/admin' -X PROPFIND -u ... -v
-> 207
```

Thus the desktop client uses the reply it gets, unaware that it is missing the external storage folder due to maintenance mode being enabled.

**Comment by guruz** [comment]
New dav point is @DeepDiver1975

**Comment by PVince81** [comment]
strange, on v10.0.0 git version I do get a 503 on "dav/files/admin" but with a strange exception:
```
HTTP/1.1 503 Service Unavailable
Date: Fri, 05 May 2017 13:30:41 GMT
Server: Apache
X-Powered-By: PHP/7.0.15
Set-Cookie: oc2d13a6a068=7fthd58lgto3tv7h61gkhutc8jl1kackroghd8p5ksorusde50l1; path=/owncloud; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
Set-Cookie: oc_sessionPassphrase=wraYiZfwyFGcrqm2HwQ7FOl5LeQ%2FMKIjNlAXYUdnEdTHhLmxYf0Sk2%2F5VA%2Bv1bjT0dYgGzvdxOSMwBoQWKblOXlvaUlF7ENE%2FxZp3FHMe7mkMsS8ISctDxj19wU9Wdq1; path=/owncloud; HttpOnly
Content-Security-Policy: default-src 'none';
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: none
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Set-Cookie: oc2d13a6a068=q17167dik7ih00u70cfe85iamrjdfoqtp18gut767596iolvta60; path=/owncloud; HttpOnly
Set-Cookie: cookie_test=test; expires=Fri, 05-May-2017 14:30:41 GMT; Max-Age=3600
Content-Length: 266
Connection: close
Content-Type: application/xml; charset=utf-8

<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\ServiceUnavailable</s:exception>
  <s:message>Error: Class 'OCA\FederatedFileSharing\AddressHandler' not found</s:message>
</d:error>
```

**Comment by PVince81** [comment]
Same result with the 10.0.0 release tarball.

Okay, I was doing a PROPFIND with Depth.
If I do a simple PROPFIND I do get a 207:

```
  % curl -D - -X PROPFIND -u admin:admin http://localhost/owncloudtest/remote.php/dav/files/admin/ 
HTTP/1.1 207 Multi-Status
Date: Fri, 05 May 2017 13:34:53 GMT
Server: Apache
X-Powered-By: PHP/7.0.15
Set-Cookie: ochyn548m31t=ou98hkhvdnkv397s1ba32vppgs6nn2j0fh887uqm4m8ervn4k3v1; path=/owncloudtest; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
Set-Cookie: oc_sessionPassphrase=Da%2BalENrdwQ7DaUMwnJRXsQGlC8iCh2iamceJagfbYpllBn0PrRxucn1mAWF6AIh%2BaTIClkCzr%2BdO69%2BAKO%2BHxqUH%2FfneY4h0goTyQKQsmOWn0GBqS6OP3tXHIGiTFMd; path=/owncloudtest; HttpOnly
Content-Security-Policy: default-src 'none';
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: none
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Set-Cookie: ochyn548m31t=bfmkm0nvfncrju9rgfgnj01alm5piqisjbbq2922doo5b012n7n1; path=/owncloudtest; HttpOnly
Set-Cookie: cookie_test=test; expires=Fri, 05-May-2017 14:34:53 GMT; Max-Age=3600
Vary: Brief,Prefer
DAV: 1, 3, extended-mkcol, access-control, calendarserver-principal-property-search, calendar-access, calendar-proxy, calendar-auto-schedule, calendar-availability, calendarserver-subscribed, oc-resource-sharing, oc-calendar-publishing, calendarserver-sharing, addressbook
Content-Length: 1988
Content-Type: application/xml; charset=utf-8

<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:cal="urn:ietf:params:xml:ns:caldav" xmlns:cs="http://calendarserver.org/ns/" xmlns:card="urn:ietf:params:xml:ns:carddav" xmlns:oc="http://owncloud.org/ns"><d:response><d:href>/owncloudtest/remote.php/dav/files/admin/</d:href><d:propstat><d:prop><d:getlastmodified>Fri, 05 May 2017 13:33:42 GMT</d:getlastmodified><d:resourcetype><d:collection/></d:resourcetype><d:quota-used-bytes>5202967</d:quota-used-bytes><d:quota-available-bytes>-3</d:quota-available-bytes><d:getetag>&quot;590c7f36d585d&quot;</d:getetag></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response><d:response><d:href>/owncloudtest/remote.php/dav/files/admin/Documents/</d:href><d:propstat><d:prop><d:getlastmodified>Fri, 05 May 2017 13:33:42 GMT</d:getlastmodified><d:resourcetype><d:collection/></d:resourcetype><d:quota-used-bytes>36227</d:quota-used-bytes><d:quota-available-bytes>-3</d:quota-available-bytes><d:getetag>&quot;590c7f36d585d&quot;</d:getetag></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response><d:response><d:href>/owncloudtest/remote.php/dav/files/admin/Photos/</d:href><d:propstat><d:prop><d:getlastmodified>Fri, 05 May 2017 13:33:42 GMT</d:getlastmodified><d:resourcetype><d:collection/></d:resourcetype><d:quota-used-bytes>678556</d:quota-used-bytes><d:quota-available-bytes>-3</d:quota-available-bytes><d:getetag>&quot;590c7f36c26b9&quot;</d:getetag></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response><d:response><d:href>/owncloudtest/remote.php/dav/files/admin/ownCloud%20Manual.pdf</d:href><d:propstat><d:prop><d:getlastmodified>Fri, 05 May 2017 13:33:42 GMT</d:getlastmodified><d:getcontentlength>4488184</d:getcontentlength><d:resourcetype/><d:getetag>&quot;727bf954856c974cfb4b9293b1e792f6&quot;</d:getetag><d:getcontenttype>application/pdf</d:getcontenttype></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response></d:multistatus>
```

**Comment by PVince81** [comment]
PR here, the maintenance mode plugin was missing in the new dav endpoint: https://github.com/owncloud/core/pull/27821

Only the first commit is needed for the fix, second one is for automated tests.

**Comment by ckamm** [comment]
@PVince81 Thank you! I've tested it and the client behaves as expected with the server patch applied.

**Comment by guruz** [comment]
@ckamm @fjf2002 The files are coming back downloaded from the server when you turn the maintenance mode off again, right?

**Comment by ckamm** [comment]
@guruz Yes, exactly, no data is lost. During maintenance mode, the client believes the files were deleted from the server and propagates that. They are restored when maintenance mode ends.

**Comment by fjf2002** [comment]
@guruz I cannot agree with ckamm. The files are not restored when server quits maintenance mode.

**Comment by guruz** [comment]
@fjf2002 Do you have more information on this? Was it files that were on the server before? Do you have logs?

**Comment by ckamm** [comment]
@fjf2002 Odd, that's the way it worked for me. It may be that it doesn't happen automatically, possibly I triggered a new sync manually. Can you check if the external data is restored when you touch a local file after the server's maintenance mode is disabled?

**Comment by fjf2002** [comment]
@guruz @ckamm Sorry I'm currently not at home and have no access to my server until weekend. (And ironically this very bug deleted my certificate for remote access.)

**Comment by guruz** [comment]
> (And ironically this very bug deleted my certificate for remote access.)

Can you clarify on that? :)

**Comment by fjf2002** [comment]
@ckamm; I tested again. Behaviour is as I remembered. The files won't come back after switching maintenance mode off on the server side and syncing manually on the desktop client side.

>Can you check if the external data is restored when you touch a local file after the server's maintenance mode is disabled?

Then the files get restored on the client-side. Interesting. But I haven't tested whether all missing files are restored or just the folder where I touched a file.

@guruz 
>Was it files that were on the server before? 

Yes. During my tests I never changed the data on the ownCloud server.

>Do you have logs?

Sorry but I'm not interested in sharing information about private files. If you still need logs: Client or server? What portions? Can we anonymize somehow?

>>(And ironically this very bug deleted my certificate for remote access.)

> Can you clarify on that? :)

Well, this is off-topic, but: My server is at home. When I'm away, I can access the server from remote with a certificate. This certificate was located on my notebook in my ownCloud sync folder. So when I was testing this bug from remote, the certificate got deleted and I was locked out from my server.

**Comment by guruz** [comment]
I'll re-open this then until it is understood by everyone that it is fixed now or other reporters come up and claim they got this too.

**Comment by ckamm** [comment]
My understanding is this:
* The underlying cause is fully resolved by owncloud/core#27821
* Not automatically restoring all files probably happens because the client believes its database is up-to-date, as the server reply has changed without an etag update. This is not a bug in itself, just a follow-up issue. In practice, this means that copies of the server data will only be redownloaded again once the affected folders are rediscovered, either by touching a file or by pressing `Ctrl-F6` while the affected folder is selected (>=2.4.0 only).
